pulp_python/app/viewsets.py

```diff
@@ -17,6 +41,42 @@ class PythonPackageContentViewSet(platform.ContentViewSet):
     queryset = python_models.PythonPackageContent.objects.all()
     serializer_class = python_serializers.PythonPackageContentSerializer
 
+    @transaction.atomic
+    def create(self, request):
+        try:
+            artifact = self.get_resource(request.data['artifact'], Artifact)
+        except KeyError:
+            raise serializers.ValidationError(detail={'artifact': _('This field is required')})
+
+        filename = request.data['filename']
+
+        # iterate through extensions since splitext does not support things like .tar.gz
+        for ext, packagetype in DIST_EXTENSIONS.items():
+            if filename.endswith(ext):
+                with tempfile.TemporaryDirectory() as td:
+                    temp_path = os.path.join(td, filename)
+                    shutil.copy2(artifact.file.path, temp_path)
+                    meta = DIST_TYPES[packagetype](temp_path)
+                    meta.packagetype = packagetype
+                    break
+        else:
+            raise ValueError(_("Filename extension is not a known python extension."))
+
+        data = parse_project_metadata(meta.__dict__)
+        data['classifiers'] = [{'name': classifier} for classifier in meta.classifiers]
+        data['packagetype'] = meta.packagetype
+        data['version'] = meta.version
+
+        # Update dictionary with user supplied metadata
+        data.update(request.data)
```

Comment:
Right now filename definitely need to be user provided in order for pkginfo to be able to parse it, should we make it possible to override other metadata? 
Old pypi use to support override, but warehouse does not...